src/parser.ts

```diff
@@ -0,0 +1,759 @@
+	f: (a: A, b: B, c: C, d: D, e: E, f: F, g: G, h: H) => I,
+) => (
+	pa: Parser<A>,
+	pb: Parser<B>,
+	pc: Parser<C>,
+	pd: Parser<D>,
+	pe: Parser<E>,
+	pf: Parser<F>,
+	pg: Parser<G>,
+	ph: Parser<H>,
+): Parser<I> =>
+	parser((a) => {
+		const resultA = pa.run(a);
+
+		if (resultA.kind === ResultKind.Err) {
+			return resultA;
+		}
+
+		const resultB = pb.run(a);
+
+		if (resultB.kind === ResultKind.Err) {
+			return resultB;
+		}
+
+		const resultC = pc.run(a);
+
+		if (resultC.kind === ResultKind.Err) {
+			return resultC;
+		}
+
+		const resultD = pd.run(a);
+
+		if (resultD.kind === ResultKind.Err) {
+			return resultD;
+		}
+
+		const resultE = pe.run(a);
+
+		if (resultE.kind === ResultKind.Err) {
+			return resultE;
+		}
+
+		const resultF = pf.run(a);
+
+		if (resultF.kind === ResultKind.Err) {
+			return resultF;
+		}
+
+		const resultG = pg.run(a);
+
+		if (resultG.kind === ResultKind.Err) {
+			return resultG;
+		}
+
+		const resultH = ph.run(a);
+
+		if (resultH.kind === ResultKind.Err) {
+			return resultH;
+		}
+
+		return ok(
+			f(
+				resultA.value,
+				resultB.value,
+				resultC.value,
+				resultD.value,
+				resultE.value,
+				resultF.value,
+				resultG.value,
+				resultH.value,
+			),
+		);
+	});
+
+/**
+ * Like `map` but applies a function that *also* returns an `Parser`. Then
+ * "unwraps" the result for you so you don't end up with `Parser<Parser<A>>`.
+ * Can be useful for stringing together multiple parsing steps, where each step
+ * relies on the result of the previous one.
+ *
+ * If the first parser fails, the function won't be called and the error will
+ * be returned instead. If the second parser fails the error for that will be
+ * returned.
+ *
+ * @param f The function to apply
+ * @param pa The parser whose result is to be passed to the function
+ * @return {Parser<B>} A new parser, built from `pa` and `f`
+ * @example
+ * type MultipleChoiceAnswer = 'a' | 'b' | 'c';
+ *
+ * const answerValueParser = (value: string): Parser<MultipleChoiceAnswer> =>
+ *   value === 'a' || value === 'b' || value === 'c'
+ *     ? succeed(value)
+ *     : fail("Needed 'a', 'b' or 'c'.");
+ *
+ * const answerParser = pipe(
+ *   fieldParser('answer', parseString),
+ *   andThen(answerValueParser),
+ * );
+ *
+ * const jsonA: unknown = JSON.parse('{ "answer": "a" }');
+ * const resultA = parse(answerParser)(jsonA); // Ok<MultipleChoiceAnswer>, 'a'
+ *
+ * const jsonB: unknown = JSON.parse('{ "answer": "d" }');
+ * const resultB = parse(answerParser)(jsonB); // Err<string>, "Needed 'a', 'b' or 'c'."
+ */
+const andThen = <A, B>(f: (a: A) => Parser<B>) => (pa: Parser<A>): Parser<B> =>
+	parser((a) => resultAndThen<string, A, B>((x) => f(x).run(a))(pa.run(a)));
+
+const oneOf = <A>(pas: Array<Parser<A>>): Parser<A> =>
+	parser((a) => {
+		const f = (
+			remPas: Array<Parser<A>>,
+			errs: string[],
+		): Result<string, A> => {
+			if (pas.length === 0) {
+				return err(errs.join(' '));
```

Comment:
Could display a more helpful message if `pas` was initially empty array (ie `errs` is empty at this stage)

Reply:
Good point. I've actually got a note to look at this function because I think I've made a bit of a mess of the base cases and error reporting.

As it happens, I don't think the `length === 0` case will _ever_ be hit unless the original array is empty, in which case it won't be hit when recursing through a larger array, in which case it probably doesn't need to live in the recursive function. Therefore, it probably makes sense to pull it into the enclosing scope and report an empty array error as you suggested.

The `errs.join(' ')` should then probably live in the `length === 1` case, because at the moment that's swallowing all prior errors and just reporting the last one.

In summary, the function should work fine, but it could report errors better.

Reply:
Indeed, I've no concerns about how the function works, just the error reporting.

Reply:
I've refactored it. `f` now only has the one base case (`length === 0`), and concats the error messages properly.

Regarding an initial empty list, I've opted to include a check at the top-level. Alternatively it could be rolled into the `length === 0` check inside `f`, but I thought it might be clearer to read if separated. What do you think?

Reply:
I agree the approach you've taken is cleaner